drivers/dac/dac_ad559x.c

```diff
@@ -0,0 +1,110 @@
+/*
+ * Copyright (c) 2023 Grinn
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#define DT_DRV_COMPAT adi_ad559x_dac
+
+#include <zephyr/drivers/dac.h>
+#include <zephyr/kernel.h>
+#include <zephyr/sys/byteorder.h>
+
+#include <zephyr/drivers/mfd/ad559x.h>
+
+#include <zephyr/logging/log.h>
+LOG_MODULE_REGISTER(dac_ad559x, CONFIG_DAC_LOG_LEVEL);
+
+#define AD559X_DAC_RESOLUTION        12
+#define AD559X_DAC_WR_POINTER        0x10
+#define AD559X_DAC_WR_MSB_BIT        BIT(15)
+#define AD559X_DAC_CHANNEL_SHIFT_VAL 12
+
+struct dac_ad559x_config {
+	const struct device *mfd_dev;
+};
+
+struct dac_ad559x_data {
+	uint8_t dac_conf;
+};
+
+static int dac_ad559x_channel_setup(const struct device *dev,
+				    const struct dac_channel_cfg *channel_cfg)
+{
+	const struct dac_ad559x_config *config = dev->config;
+	struct dac_ad559x_data *data = dev->data;
+
+	if (channel_cfg->channel_id >= AD559X_PIN_MAX) {
+		LOG_ERR("Invalid channel number %d", channel_cfg->channel_id);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->resolution != AD559X_DAC_RESOLUTION) {
+		LOG_ERR("Invalid resolution %d", channel_cfg->resolution);
+		return -EINVAL;
+	}
+
+	data->dac_conf |= BIT(channel_cfg->channel_id);
+
+	return mfd_ad559x_write_reg(config->mfd_dev, AD559X_REG_LDAC_EN, data->dac_conf);
+}
+
+static int dac_ad559x_write_value(const struct device *dev, uint8_t channel, uint32_t value)
+{
+	const struct dac_ad559x_config *config = dev->config;
+	uint16_t msg;
+
+	if (channel >= AD559X_PIN_MAX) {
+		LOG_ERR("Invalid channel number %d", channel);
+		return -EINVAL;
+	}
+
+	if (value >= (1 << AD559X_DAC_RESOLUTION)) {
+		LOG_ERR("Value %d out of range", value);
+		return -EINVAL;
+	}
+
+	if (mfd_ad559x_has_pointer_byte_map(config->mfd_dev)) {
```

Comment:
Can we detect the bus somehow in the `mfd_ad559x` driver? I would like to have a generic i.e `mfd_ad559x_write_reg` function that would work for both I2C and SPI variants.

Reply:
`write_reg` and `write_raw` are defined in the [mfd_ad559x_transfer_function](https://github.com/jeppenodgaard/zephyr/blob/50393e142f5e65a402ffba62fb3e11059ac5f84a/drivers/mfd/mfd_ad559x.h#L35) so it works for both SPI and I2C.

The register mapping is different (for some reason) for these two ICs and the registers does not always work in the same way. See e.g.:
https://github.com/jeppenodgaard/zephyr/blob/50393e142f5e65a402ffba62fb3e11059ac5f84a/drivers/adc/adc_ad559x.c#L113

I do not think it is possible to make a struct/union mapping.

Reply:
@bbilas did I misunderstand your question / suggestion?

Reply:
Okay for me.